services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/service/MongoDbBasedRegistrationService.java

```diff
@@ -0,0 +1,369 @@
+import com.mongodb.MongoException;
+
+import io.opentracing.Span;
+import io.vertx.core.AbstractVerticle;
+import io.vertx.core.Future;
+import io.vertx.core.Promise;
+import io.vertx.core.json.JsonArray;
+import io.vertx.core.json.JsonObject;
+import io.vertx.ext.mongo.IndexOptions;
+import io.vertx.ext.mongo.MongoClient;
+import io.vertx.ext.mongo.MongoClientDeleteResult;
+import io.vertx.ext.mongo.MongoClientUpdateResult;
+
+/**
+ * This is an implementation of the device registration service and the device management service where data 
+ * is stored in a mongodb database.
+ *
+ * @see <a href="https://www.eclipse.org/hono/docs/api/device-registration/">Device Registration API</a>
+ * @see <a href="https://www.eclipse.org/hono/docs/api/management/">Device Registry Management API</a>
+ */
+@Component
+@Qualifier("serviceImpl")
+@ConditionalOnProperty(name = "hono.app.type", havingValue = "mongodb", matchIfMissing = true)
+public class MongoDbBasedRegistrationService extends AbstractVerticle
+        implements DeviceManagementService, RegistrationService {
+
+    private static final Logger log = LoggerFactory.getLogger(MongoDbBasedRegistrationService.class);
+    private MongoClient mongoClient;
+    private MongoDbBasedRegistrationConfigProperties config;
+    private MongoDbCallExecutor mongoDbCallExecutor;
+
+    /**
+     * Registration service, based on {@link AbstractRegistrationService}.
+     * <p>
+     * This helps work around Java's inability to inherit from multiple base classes. We create a new Registration
+     * service, overriding the implementation of {@link AbstractRegistrationService} with the implementation of our
+     * {@link MongoDbBasedRegistrationService#getDevice(String, String, Span)}.
+     */
+    private final AbstractRegistrationService registrationService = new AbstractRegistrationService() {
+
+        @Override
+        public Future<RegistrationResult> getDevice(final String tenantId, final String deviceId, final Span span) {
+            return MongoDbBasedRegistrationService.this.getDevice(tenantId, deviceId, span);
+        }
+
+        @Override
+        public Future<JsonArray> resolveGroupMembers(final String tenantId, final JsonArray viaGroups, final Span span) {
+            return MongoDbBasedRegistrationService.this.resolveGroupMembers(tenantId, viaGroups);
+        }
+    };
+
+    /**
+     * Creates an instance of the {@link MongoDbCallExecutor}.
+     *
+     * @param mongoDbCallExecutor An instance of the mongoDbCallExecutor.
+     */
+    @Autowired
+    public void setExecutor(final MongoDbCallExecutor mongoDbCallExecutor) {
+        this.mongoDbCallExecutor = mongoDbCallExecutor;
+        this.mongoClient = this.mongoDbCallExecutor.getMongoClient();
+    }
+
+    @Autowired
+    public void setConfig(final MongoDbBasedRegistrationConfigProperties config) {
+        this.config = config;
+    }
+
+    public MongoDbBasedRegistrationConfigProperties getConfig() {
```

Comment:
is this going to be used?

Reply:
This method `getConfig()`is already being used instead of directly using the instance variable `config`.

Reply:
But that is not necessary, is it?
The question is: will any other class be required to get access to this service's configuration using this method? If not (which I assume) we should remove the method, shouldn't we?

Reply:
It makes sense and I will remove it.